File: app/schemas/ml_command_center.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class ActivePermitInput(BaseModel):
    permit_id: str
    permit_type: str
    status: str


class SensorAssessmentInput(BaseModel):
    factory: str
    region: str
    zone_id: str

    shift: Literal["day", "night"]
    workers: int = Field(ge=0)
    experience_level: Literal["junior", "senior"]
    training: Literal["yes", "no"]

    # Accident-risk model fields
    temperature: float = Field(ge=-20, le=100)
    pressure: float
    humidity: float
    vibration: float
    machine_speed: float
    equipment_age: int = Field(ge=0)
    service_days: int = Field(ge=0)
    gas: float
    sparks: int = Field(ge=0)
    alarm: Literal["on", "off"]

    # Machine-anomaly model fields
    motor_current: float
    rpm: float
    ambient_humidity: float
    tool_wear: float
    coolant_flow_rate: float
    voltage_fluctuation_percent: float
    operator_experience_years: float

    timestamp: datetime
    active_permit: ActivePermitInput | None = None
    
class PredictionResponse(BaseModel):
    model_name: str
    predicted_class: int
    predicted_event: str
    accident_probability: float


class AnomalyResponse(BaseModel):
    available: bool
    is_anomaly: bool
    raw_anomaly_score: float | None
    anomaly_component_score: float


class TriggeredRuleResponse(BaseModel):
    rule_id: str
    severity: str
    message: str


class RuleDetectionResponse(BaseModel):
    triggered: bool
    triggered_rule_count: int
    rule_score: float
    triggered_rules: list[TriggeredRuleResponse]


class PermitConflictResponse(BaseModel):
    conflict_detected: bool
    permit_id: str | None
    permit_type: str | None
    reason: str
    recommended_action: str


class WorkerExposureResponse(BaseModel):
    workers_present: int
    workers_exposed: int
    evacuation_required: bool
    training_status: str
    experience_level: str


class CompoundAssessmentResponse(BaseModel):
    ml_score: float
    anomaly_score: float
    rule_score: float
    operational_score: float
    compound_risk_score: float
    risk_level: str
    final_decision: str
    contributing_factors: list[str]


class SensorIngestionResponse(BaseModel):
    status: str
    factory: str
    region: str
    zone_id: str
    timestamp: datetime


class CommandCenterAssessmentResponse(BaseModel):
    sensor_ingestion: SensorIngestionResponse
    ml_prediction: PredictionResponse
    anomaly_detection: AnomalyResponse
    rule_detection: RuleDetectionResponse
    permit_conflict: PermitConflictResponse
    worker_exposure: WorkerExposureResponse
    compound_assessment: CompoundAssessmentResponse

File: frontend/src/types/mlCommandCenter.ts
export interface ActivePermit {
  permit_id: string;
  permit_type: string;
  status: string;
}

export interface SensorAssessmentRequest {
  factory: string;
  region: string;
  zone_id: string;

  shift: "day" | "night";
  workers: number;
  experience_level: "junior" | "senior";
  training: "yes" | "no";

  temperature: number;
  pressure: number;
  humidity: number;
  vibration: number;
  machine_speed: number;
  equipment_age: number;
  service_days: number;
  gas: number;
  sparks: number;
  alarm: "on" | "off";

  motor_current: number;
  rpm: number;
  ambient_humidity: number;
  tool_wear: number;
  coolant_flow_rate: number;
  voltage_fluctuation_percent: number;
  operator_experience_years: number;

  timestamp: string;
  active_permit: ActivePermit | null;
}

export interface TriggeredRule {
  rule_id: string;
  severity: string;
  message: string;
}

export interface CommandCenterAssessment {
  sensor_ingestion: {
    status: string;
    factory: string;
    region: string;
    zone_id: string;
    timestamp: string;
  };

  ml_prediction: {
    model_name: string;
    predicted_class: number;
    predicted_event: string;
    accident_probability: number;
  };

  anomaly_detection: {
    available: boolean;
    is_anomaly: boolean;
    raw_anomaly_score: number | null;
    anomaly_component_score: number;
  };

  rule_detection: {
    triggered: boolean;
    triggered_rule_count: number;
    rule_score: number;
    triggered_rules: TriggeredRule[];
  };

  permit_conflict: {
    conflict_detected: boolean;
    permit_id: string | null;
    permit_type: string | null;
    reason: string;
    recommended_action: string;
  };

  worker_exposure: {
    workers_present: number;
    workers_exposed: number;
    evacuation_required: boolean;
    training_status: string;
    experience_level: string;
  };

  compound_assessment: {
    ml_score: number;
    anomaly_score: number;
    rule_score: number;
    operational_score: number;
    compound_risk_score: number;
    risk_level: string;
    final_decision: string;
    contributing_factors: string[];
  };
}

export interface RegionOverview {
  factory: string;
  region: string;
  zone_id: string;
  timestamp: string;
  ml_probability: number;
  is_anomaly: boolean;
  compound_risk_score: number;
  risk_level: string;
  workers_exposed: number;
  final_decision: string;
}

export interface CommandCenterOverview {
  summary: {
    total_regions: number;
    safe_regions: number;
    alert_regions: number;
    high_regions: number;
    critical_regions: number;
    workers_exposed: number;
  };

  safe_regions: RegionOverview[];
  alert_regions: RegionOverview[];
  regions: RegionOverview[];
}

export interface ModelMetrics {
  available: boolean;
  model_name: string | null;
  accuracy?: number;
  precision?: number;
  recall?: number;
  f1_score?: number;
  roc_auc?: number;
  pr_auc?: number;
  false_negative_rate?: number;

  confusion_matrix?: {
    true_negative: number;
    false_positive: number;
    false_negative: number;
    true_positive: number;
  };
}